### ISSUE TITLE
pheatmap annotation error

### ISSUE THREAD

**Issue by Tristan-Philippe** [STAT540-UBC/Discussion#215]
Hello,

I don't know how much I should post on here considering this is for the analysis assignment. I am having issues adding annotations to my heatmap (using pheatmap). It works fine without the annotations, but when I add the annotation:

annotation_col = des[, c("organism_part", "cell_type", "batch", "age")],

I get the following error: "Error in cut.default(a, breaks = 100) : 'x' must be numeric".
I based my code on the code used in seminar 6 which is similar to what I read on the internet. I noticed the data was a matrix in the homework so I tried to coerce my data (which is a data.frame) as a matrix, which seemed to work for the time being, but when I tried to knit everything the same error was back and now it won't go away. I even tried to make it a matrix and it is giving me the same weird error again.
My version of R, R studio, and all of my packages are up to date. I tried removing and re-installing pheatmap.

Any suggestions would be appreciated.

Thanks,
Tristan

**Comment by eneskemalergin** [comment]
Hello @Tristan-Philippe  
I got the same issue recently and there were several things that I got to fix:
- Make sure you have a matrix for the expression data. 
- annotation data.frame should have samples as the index of the dataframe. 
- make sure to change the type of annotation data.frame to as.data.frame(des) (This is stupid I know, but pheatmap's annotation taker cannot handle types other than data.frame, like tibbles)

Hope it helps

**Comment by ppavlidis** [comment]
When I get this error, it has meant my annotation input was empty.

pheatmap knows how to make annotations from factors, characters or, as a last resort, numbers, which is where `cut` comes into play - but there it simply assumes that the values are numbers without checking - empty inputs will end up there, causing the error. To me, that's a bug in pheatmap. 

(See https://github.com/raivokolde/pheatmap/blob/master/R/pheatmap.r around line 272)

If @eneskemalergin's advice doesn't get you anywhere (Does it? If so my diagnosis is off-base), my approach to narrow this down would be to try a single column to annotate.

In other words ensure that `des[, c("organism_part", "cell_type", "batch", "age")]` is doing what you think it does.

If you solve it please update the thread with the solution!

**Comment by sepkamal** [comment]
Not sure if it's exactly the same problem, but I got the same error. 

Took me a while to figure out, but it turned out to be a problem with the format of the metadata dataframe. The sample ID's were stored in the column "samples", but actually needed to be set as the rownames.

Hope this helps,
Sep

**Comment by Tristan-Philippe** [comment]
Hello,

@eneskemalergin Thanks for clarifying those three points. I made sure those three conditions were met, but it still didn't work.

As Dr. Pavlidis mentioned, I thought I would check that it's not empty and make a data.frame from it. Then added an index to the data.frame since it lost the rownames. Then used that new data.frame (I called annotation) and it works. The code is below:

annotation = data.frame(des[, c("organism_part", "cell_type", "batch", "age")])
rownames(annotation) <- colnames(dat)
pheatmap(as.matrix(logdat), cluster_rows = FALSE, cluster_cols = FALSE, show_rownames = FALSE, main = "Heatmap for GSE60019", fontsize = 5,
         annotation_col = annotation,
         annotation_colors = list(cell_type, organism_part, batch, age))

Note that I specified the colours I wanted before, and turned off clustering and rownames for this plot. I also made the font size 5 for everything so that the annotation labels would fit in the figure.

Seems like I am just having a strange bug. I guess it didn't feel like extracting the columns so thought it was empty, but sometimes it feels like it? At least it works now.

Thanks for your help,
Tristan

**Comment by eneskemalergin** [comment]
@Tristan-Philippe 

Can you copy here the output of the following: 

```R
class(annotation)
```

**Comment by Tristan-Philippe** [comment]
Sure, in case I wasn't clear using "annotation" worked.
[1] "data.frame"

I asked if des was a dataframe before (using is.data.frame) and it said true, so I thought it was fine. But when checking "class(des)" the output is " [1] "tbl_df"     "tbl"        "data.frame" ". I then tried to rewrite des as a data.frame "des <- data.frame(des)" and wrote what I had before as the annotation : "des[, c("organism_part", "cell_type", "batch", "age")]" and it seems to work now too. That's probably what I did at one point, but it didn't work so I deleted it. While I was playing with it des was still only a data.frame so it worked once I coerced the expression data into a matrix leading me to think that coercing the expression data into a matrix was the only issue. I guess next time I will check the class rather than ask true or false questions.

I guess pheatmap cannot handle an object that is more than just a data.frame.

Tristan

**Comment by eneskemalergin** [comment]
Yeah, that's what I was saying. Believe me It was really frustrating for me to find out that it has to be only data.frame after trying to solve the issue for myself.

**Comment by Tristan-Philippe** [comment]
Ok thanks that makes sense now. I am a novice at this so I didn't realize objects could be more than one thing.

**Comment by ppavlidis** [comment]
Thanks everybody, I think we figured it out. Just to clarify (hopefully correctly):

The problem is even if tibbles pretend to be data frames, they don't have row names. The developer of tibbles is anti-row names (religiously). I believe this is fundamentally the issue.

Because I forgot the other cause of this error I have seen, which is what @eneskemalergin said about the row names of your `des` having to match the names in your data. Because pheatmap does this (where `mat` is the data):

`  # Select only the ones present in the matrix`
` annotation_col = annotation_col[colnames(mat), , drop = F]` 

(similar for annotation_row)

This *is* a bug in pheatmap that it doesn't bother checking whether `annotation_col` is is now empty. Compounded by this rather tricky behaviour of tibbles.
